README.md
# Nifty Nav
Nifty Nav is a quick and easy navigation plugin using Sass and jQuery. It makes
making navigation systems a little better. Built by the awesome people over at [factor1](http://factor1studios.com).

## Getting Started

##### 1. Install Nifty Nav with `npm` or `bower`

```
npm install nifty-nav --save
## OR
bower install nifty-nav --save
```

##### 2. Import the Nifty Nav styles & settings in your `.scss` file

To use the default settings:

```scss
@import nifty-nav/settings;
@import nifty-nav/nifty-nav;
```

To customize the settings, copy `settings.scss` to your project folder and include
it before `nifty-nav`. This example moves `settings.scss` to a folder called
`settings` and renames it to `nifty-nav-settings`:

```scss
@import settings/nifty-nav-settings;
@import nifty-nav/nifty-nav;
```

##### 3. Include the `nifty-nav.js`
Include the `nifty-nav.js` in your build either inline or within a Gulp or Grunt
task.

### Quick Setup
To use Nifty Nav without compiling and with the default settings, clone the Nifty Nav repo and use
`nifty-nav.min.css` found in `dist/css/` and `nifty-nav.min.js` found in `dist/js/`.

```html
<html>
  <head>
    <link rel="stylesheet" href="path/to/file/nifty-nav.min.css">
  </head>
  <body>
   ...
   <script src="path/to/file/nifty-nav.min.js"></script>
  </body>
</html>
```

## Adding The Hamburger
To add the hamburger to your site:

```html
<a id="nifty-nav-toggle"><span></span></a>
```

## Nifty Nav Options/Settings
Nifty Nav has five `jQuery` settings, and some `scss` settings.

- `subMenus` controls if Nifty Nav should allow toggling of sub-menus. Default: `false`.
- `subMenuParentLink` controls if a parent menu navigation link should be followed when clicked. Default: `false`.
- `mask` controls if Nifty Nav should show the mask that covers the page. Default: `true`.
- `itemClickClose` controls if Nifty Nav should close when a nav item with the class of `nifty-nav-item`
is clicked. Default: `true`.
- `panelPosition` lets you control the `display:` value of the `.nifty-panel` element. Default: `absolute`.

> **NOTE:**  When using `panelPosition: 'fixed'` a class of `.nifty-lock` is added
> to the `<body>`. This is to prevent the page from scrolling and allow the user to scroll
> through nav items instead. To change this functionality simply override the `.nifty-lock` class.


`Scss` settings control visual aspects of nifty nav:

- `$header-index` sets the z-index of the `<header>` element
- `$panel-color` sets the color of the dropdown.
- `$panel-top-distance` sets the panel distance from the top of the viewport.
- `$panel-width` sets the panel width.
- `$panel-index` sets the z-index of the Nifty Nav panel.
- `$mask-color` sets the color and opacity of the mask.
- `$mask-index` sets the z-index of the mask.
- `$hamburger-color` sets the hamburger color.
- `$hamburger-color-active` sets the color of the hamburger when the panel is open.
- `$hamburger-position-top` sets the distance from the top of the containing element.
- `$hamburger-position-x` sets the distance from the left/right of the containing element.
- `$hamburger-position-left` - default is `true`. If false, the hamburger floats to the right.

## Initializing Nifty Nav
To use Nifty Nav, initialize it at the bottom of your page before the closing `</body>` or in an external JavaScript file (recommended).

```js
// Nifty Nav with default settings
$(document).ready(function(){
	$('#nifty-nav-toggle').niftyNav();
});

// Nifty Nav without the mask and with sub-menu toggles
$(document).ready(function(){
  niftyNav({
    mask: false,
    subMenus: true
  });
});
```


# Issues / Bugs / Contributions#
If you have feedback, find a bug, or want to make contributions, please don't hesitate
to open an issue or make a pull request.


changelog.txt
/////////////////////////
—— F1 Nav Change Log ——
////////////////////////


VERSION 3.0.0 05/17/2017
___________________________
Brings versions ~2.3.0 to 3.0.0 since it was a breaking change.

VERSION 2.3.1 04/17/2017
___________________________
Hotfix to add class to correct LI instead of a per issue #21

VERSION 2.30 04/17/2017
___________________________
Makes the niftyNav an actual jQuery plugin.

VERSION 2.2.2 12/22/2016
___________________________
Fixes toggle clickable area.

VERSION 2.2.1 12/19/2016
___________________________
Makes clickable area larger, updates dev devDependencies

VERSION 2.1.4 04/23/2016
___________________________
Fixes a "bug" that breaks parent nav items site wide. Thanks to @bebaps for pointing this out.

VERSION 2.1.3 03/21/2016
___________________________
Adds a `subMenuParentLink` option to allow/block the browser from following parent navigation item links.

VERSION 1.5 11/3/2015
___________________________
Version 1.5.0 adds sub menu toggles, as well as a complete overhaul of the file structure in preparation of bumping to 2.0.0 which will include a bower package for Nifty Nav.
Note that most if not all of the class names have changed to avoid conflicts with Wordpress or other jQuery based plugins.


VERSION 1.4 12/19/2014
———————————————————————————
Version 1.4 is siq. We added transitions to the mask (finally!) so that it looks nice and pretty on open, and if you’re using a one page layout, looks just as slick.

OH HEY! Also, it’s good to not repeat yourself, so we threw out some lines of code by defining the functions first, so it makes the code look a bit cleaner. I’m sure there’s more work to be done with that…

VERSION 1.3+ Now lives on GitHub, no longer bitbucket.
___________________________

VERSION 1.3 12/17/2014
———————————————————————————
Version 1.3 adds the ability to close the nav when clicking on the navigation link. The class is applied to the LI with a class of ‘nav-item’.

VERSION 1.2 12/16/2014
———————————————————————————
Version 1.2 adds some new functionality and eye candy. We ditched font awesome for the ‘hamburger’ icon and use <span></span> to allow for an animation/transform into an X for closing the nav. This includes some big changes to the jQuery as well as a handful of CSS properties.

The jQuery now counts the clicks from the user on both the mask and the icon, to apply or remove the mask.


VERSION 1.1 10/22/2014
———————————————————————————
Version 1.1 improves upon the initial 1.0 where all the nav did was push down the body, creates a mask that overlays the body to darken the content and detect clicks to close the nav.  (those were the days).


gulpfile.js
// Include gulp
var gulp = require('gulp');

// Include Our Plugins
var clean = require('gulp-clean');
var jshint = require('gulp-jshint');
var sass = require('gulp-sass');
var uglify = require('gulp-uglify');
var rename = require('gulp-rename');
var minifyCss = require('gulp-minify-css');

// Clean Dist Folder
gulp.task('clean', function() {
	return gulp.src('./dist/***/*', {read: false})
	.pipe(clean({force: true}));
});

// Lint Task
gulp.task('lint', function() {
    return gulp.src('./src/js/*.js')
        .pipe(jshint())
        .pipe(jshint.reporter('default'));
});

// Compile Our Sass and minify it
gulp.task('sass', function() {
    return gulp.src('./src/scss/*.scss')
        .pipe(sass())
        .pipe(rename({
	        suffix: '.min'
        }))
        .pipe(minifyCss({
		    compatibility: 'ie8',
		    aggressiveMerging: false,
		    processImport: true
	    }))
	    .pipe(gulp.dest('./dist/css'));
});

// Concatenate & Minify JS
gulp.task('scripts', function() {
    return gulp.src('./src/js/*.js')
        .pipe(rename({
	        suffix: '.min'
        }))
        .pipe(uglify())
        .pipe(gulp.dest('./dist/js'));
});

// Watch Files For Changes
gulp.task('watch', function() {
    gulp.watch('./src/js/*.js', ['lint', 'scripts']);
    gulp.watch('./src/scss/*.scss', ['sass']);
    //gulp.watch('css/src/*.css', ['minify-css']);
});


// Default Task
gulp.task('default', ['clean','lint', 'sass', 'scripts', 'watch']);


src/js/new.js
// Proposed Changes To Nify Nav


  var niftyNav = function(options){
    $nifty_toggle = $('#nifty-nav-toggle');
    $nifty_panel = $('.nifty-panel');
    $nifty_nav_item = $('.nifty-nav-item');
    $nifty_parent = $('.nifty-panel ul li');

    var settings = $.extend({
      // These are the defaults.
      sub_menus: false,
      use_mask: true,
      item_click_close: true
    }, options );

    sub_menus         = settings.sub_menus;
    use_mask          = settings.use_mask;
    item_click_close  = settings.item_click_close;


    // Core Nifty Nav Functions
    niftyRemove = function(){
      $('.nifty-mask').remove();
    };

    niftyUnmask = function(){
      $('.nifty-mask').animate({opacity: 0.0});
      setTimeout(niftyRemove, 800);
    };

    // on nifty nav toggle click
    $nifty_toggle.click(function(){
      var $this = $(this);
      $this.toggleClass('nifty-active');
      $nifty_panel.slideToggle(500).css('position','absolute');

      if( use_mask === true){
        // if a mask exists
        if( $('.nifty-mask').length > 0 ){
          niftyUnmask();
        } else{
          // if no mask exists
          $('body').append('<div class="nifty-mask"></div>');
          $('.nifty-mask').animate({opacity: 1.0});

          // if a user clicks on the mask
          $('.nifty-mask').click(function(){
            $nifty_panel.slideUp(500);
            niftyUnmask();
            $nifty_toggle.removeClass('nifty-active');
          });
        }
      }
    });

    // close nifty nav on nav item click
    if( item_click_close === true ){
      $nifty_nav_item.click(function(){
        $nifty_panel.slideUp(500);
        niftyUnmask();
        $nifty_toggle.removeClass('nifty-active');
      });
    }

    // if sub menus are enabled
    if( sub_menus === true ){
      var $nifty_parent_active;
      $nifty_parent.click(function(){
        $nifty_parent_active = $(this);
        $nifty_parent_active.find('.sub-menu').slideToggle();
        $nifty_parent_active.find('a').toggleClass('nifty-menu-opened');
      });
    }

  };


src/js/f1-nav.js
//////////////////////////////////////////
//										//
//			 NIFTY NAV					//
//	 BY ERIC STOUT / FACTOR1 STUDIOS	//
//	  HTTP://FACTOR1STUDIOS.COM			//
//										//
//////////////////////////////////////////

// function we use to remove the mask
var niftyRemove = function(){ 
	$('.nifty-mask').remove(); 
};

var niftyUnMaskIt = function(){
	$('.nifty-mask').animate({opacity:0.0});	
	setTimeout(niftyRemove, 800);
};

// setting up our variables for submenu toggles
var mobileNavItem = $('.nifty-panel ul li');
var $mobileClicked;

// when the document is ready lets go!
$(document).ready(function(){

	i = 0;

	// listen for the click on the menu icon
	$('#nifty-nav-toggle').click(function(){ 
		
		// Add 1 to the count
		i++;
		 
		$('#nifty-nav-toggle').toggleClass('nifty-active');
		
		// toggle the nav up/down
		$('.nifty-panel').slideToggle(500).css("position", "absolute"); 
	
		// Check if the counter is even or odd
		var isEven = function(clickCounter){
			return (clickCounter%2 === 0) ? true : false;
		};
	
		if (isEven(i) === false){
			
			//lets create a nice mask to dim the content											
			$('body').append('<div class="nifty-mask"></div>'); 
			$('.nifty-mask').animate({opacity:1.0});	
		} else {
			niftyUnMaskIt();
		}
		
		// or alternatively you can click the mask and make that shit go away.
		$('.nifty-mask').click(function(){
			
			//slide up the nav panel
			$('.nifty-panel').slideUp(500);
			
			//remove the mask
			niftyUnMaskIt();
			
			// remove the active class on the hamburger mmmm hamburger
			$('#nifty-nav-toggle').removeClass('nifty-active');
			
			i++;  // Add 1 to the count
		});
		
	});
	// close up our listener for nav toggle

	

	/*
	Oh snap! You want to close the menu when clicking a nav-item?
	We got you... this is usefull when your site is a single page layout
	and you want to jump down to an anchor tag instead of a different page.	
	*/
	$('.nifty-nav-item').click(function(){ 
		
		//slide up our panel
		$('.nifty-panel').slideUp(500);
		niftyUnMaskIt(); 
		$('#nifty-nav-toggle').removeClass('nifty-active');
		i = 0;  // reset the counter
	});


	/*
	Listen for any menu items that have children and allow them
	to toggle open when clicked. This is by default setup for 
	Wordpress menus, however if you're not using Wordpress just 
	make sure your classes and structure matches.	
	*/
	mobileNavItem.on('click', function(){
		$mobileClicked = $(this);
		$mobileClicked.find('.sub-menu').slideToggle();
		$mobileClicked.find('a').toggleClass('menu-opened');
	});
	
});

f1-nav-development.js
////////////////////////////////////////////////////////////////////////
///////////------Factor1 Nav------//////////////////////////////////////
// a jquery tool by Eric Stout (@buschschwick) and Matt Adams (@mattada) 
// version 1.3 ////////////////////////////////////////////////////////
// http://factor1studios.com  ////////////////////////////////////////
//////////////////////////////////////////////////////////////////////
jQuery(document).ready(function($) {
	$(document).ready(function(){
		var i = 0;
		$('#nav-toggle').click(function(){ //listen for the click on the menu icon
			i++; // Add 1 to the count
			$('#nav-toggle').toggleClass('active');
			$('.panel').slideToggle(500).css("position", "fixed"); //toggle the nav up/down
			
			// Check if the counter is even or odd
			var isEven = function(clickCounter){
				return (clickCounter%2 == 0) ? true : false;
			};
			
			if (isEven(i) == false){											
				$('body').append('<div class="mask"></div>'); //lets create a nice mask to dim the content	
			} else {
				$('.mask').remove(); //removes the mask
			}

	//or alternatively you can click the mask and make that shit go away.
		$('.mask').click(function(){ //listen for the click on the mask
			$('.panel').slideUp(500); //pull up the nav
			$('.mask').remove(); //take off that mask
			$('#nav-toggle').removeClass('active'); //mmm hamburger
			i++;  // Add 1 to the count
			});
			
		$('.nav-item').click(function(){ //oh snap, you want to close the nav on a link click?
			$('.panel').slideUp(500); //pull up the nav
			$('.mask').remove(); //take off that mask
			$('#nav-toggle').removeClass('active');
			i = 0;  // reset the counter
			});
		});//close up everything
	});
});

src/js/nifty-nav.js
/* Nifty Nav
* Author: Eric Stout / Factor1
* http://factor1studios.com
* Repo: https://github.com/factor1/nifty-nav
* Version: 2.3.1
*/

(function ($) {

  $.fn.niftyNav = function(options){
    $nifty_toggle = $('#nifty-nav-toggle');
    $nifty_panel = $('.nifty-panel');
    $nifty_nav_item = $('.nifty-nav-item');
    $nifty_parent = $('.nifty-panel ul li');

    var settings = $.extend({
      // These are the defaults.
      subMenus: false,
      mask: true,
      itemClickClose: true,
      panelPosition:  'absolute',
      subMenuParentLink:  false
    }, options );

    subMenus          = settings.subMenus;
    mask              = settings.mask;
    itemClickClose    = settings.itemClickClose;
    panelPosition     = settings.panelPosition;
    subMenuParentLink = settings.subMenuParentLink;


    // Core Nifty Nav Functions
    niftyRemove = function(){
      $('.nifty-mask').remove();
    };

    niftyUnmask = function(){
      $('.nifty-mask').animate({opacity: 0.0});
      setTimeout(niftyRemove, 800);
    };

    // on nifty nav toggle click
    $nifty_toggle.click(function(){
      var $this = $(this);
      $this.toggleClass('nifty-active');
      $nifty_panel.slideToggle(500).css('position',panelPosition);

      // if panelPosition is fixed
      if( panelPosition == 'fixed' ){
        $('body').toggleClass('nifty-lock');
      }

      if( mask === true){
        // if a mask exists
        if( $('.nifty-mask').length > 0 ){
          niftyUnmask();
        } else{
          // if no mask exists
          $('body').append('<div class="nifty-mask"></div>');
          $('.nifty-mask').animate({opacity: 1.0});

          // if a user clicks on the mask
          $('.nifty-mask').click(function(){
            $nifty_panel.slideUp(500);
            niftyUnmask();
            $nifty_toggle.removeClass('nifty-active');

            // if panelPosition is fixed
            if( panelPosition == 'fixed' ){
              $('body').removeClass('nifty-lock');
            }
          });
        }
      }
    });

    // close nifty nav on nav item click
    if( itemClickClose === true ){
      $nifty_nav_item.click(function(){
        $nifty_panel.slideUp(500);
        niftyUnmask();
        $nifty_toggle.removeClass('nifty-active');

        // if panelPosition is fixed
        if( panelPosition == 'fixed' ){
          $('body').removeClass('nifty-lock');
        }

      });
    }

    // if sub menus are enabled
    if( subMenus === true ){
      var $nifty_parent_active;
      // if subMenuParentLink is false
      if( subMenuParentLink === false ){
        $('.nifty-panel .menu-item-has-children > a').click(function(event){
          event.preventDefault();
        });
      }

      $nifty_parent.click(function(){
        $nifty_parent_active = $(this);
        $nifty_parent_active.find('.sub-menu').slideToggle();
        $nifty_parent_active.toggleClass('nifty-menu-opened');
      });
    }

  };

}( jQuery ));
